File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isClosing(char ch) {
        return ch == ')' || ch == '}' || ch == ']';
    }

    bool isMatched(char open, char close) {
        return (open == '(' && close == ')') ||
               (open == '{' && close == '}') ||
               (open == '[' && close == ']');
    }

    bool isValid(string s) {
        stack<char> st;

        for(char ch : s) {
            if(isClosing(ch)) {
                if(st.empty()) return false;

                char open = st.top();
                st.pop();

                if(!isMatched(open, ch))
                    return false;
            }
            else {
                st.push(ch);
            }
        }

        return st.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    boolean isOpenning(char curr) {
        return curr == '(' || curr == '{' || curr == '[';
    }

    boolean isMatched(char open, char close) {
        return (open == '(' && close == ')') ||
                (open == '{' && close == '}') ||
                (open == '[' && close == ']');
    }

    public boolean isValid(String s) {
        int n = s.length();
        
        if(n == 1) return false;

        Stack<Character> st = new Stack<>();

        for(int i=0;i<n;i++) {
            char curr = s.charAt(i);

            if(isOpenning(curr)) {
                st.add(curr);
            }
            else {
                if (st.isEmpty()) return false;

                char popped = st.pop();
                if(!isMatched(popped, curr)) return false;
            }
        }

        return st.isEmpty();
    }
}